File: legacy/artifacts/src/parsers/BufferWrapper.ts
/* eslint-disable no-bitwise */

export class BufferWrapper {
  buffer: Buffer;

  cursor: number;

  debug: boolean;

  constructor(buffer: Buffer, debug: boolean = false) {
    this.buffer = buffer;
    this.cursor = 0;
    this.debug = debug;
  }

  seek(offset: number) {
    this.cursor = offset;
  }

  skip(length: number) {
    this.cursor += length;
  }

  readU8() {
    this.logGroup('readU8');
    this.logDebug(`cursor: ${this.cursor}`);

    const val = this.buffer[this.cursor];
    this.logDebug(`value: ${val}`);

    this.cursor += 1;

    this.logGroupEnd();
    return val;
  }

  readS8() {
    this.logGroup('readS8');
    this.logDebug(`cursor: ${this.cursor}`);

    const val = this.buffer.readInt8(this.cursor);
    this.logDebug(`value: ${val}`);

    this.cursor += 1;

    this.logGroupEnd();
    return val;
  }

  readU16() {
    this.logGroup('readU16');
    this.logDebug(`cursor: ${this.cursor}`);

    const val = this.buffer.readUInt16LE(this.cursor);
    this.logDebug(`value: ${val}`);

    this.cursor += 2;

    this.logGroupEnd();
    return val;
  }

  readS32() {
    this.logGroup('readS32');
    this.logDebug(`cursor: ${this.cursor}`);

    const val = this.buffer.readInt32LE(this.cursor);
    this.logDebug(`value: ${val}`);

    this.cursor += 4;

    this.logGroupEnd();
    return val;
  }

  readU32(): number {
    this.logGroup('readU32');
    this.logDebug(`cursor: ${this.cursor}`);

    const val = this.buffer.readUInt32LE(this.cursor);
    this.logDebug(`value: ${val} 0x${val.toString(16)}`);

    this.cursor += 4;

    this.logGroupEnd();
    return val;
  }

  readU32BE(): number {
    this.logGroup('readU32BE');
    this.logDebug(`cursor: ${this.cursor}`);

    const val = this.buffer.readUInt32BE(this.cursor);
    this.logDebug(`value: ${val} 0x${val.toString(16)}`);

    this.cursor += 4;

    this.logGroupEnd();
    return val;
  }

  readU64(): bigint {
    this.logGroup('readU64');
    this.logDebug(`cursor: ${this.cursor}`);

    const low = this.buffer.readUInt32LE(this.cursor);
    const high = this.buffer.readUInt32LE(this.cursor + 4);
    const val = (BigInt(high) << 32n) | BigInt(low);
    this.logDebug(`value: ${val} 0x${val.toString(16)}`);

    this.cursor += 8;

    this.logGroupEnd();
    return val;
  }

  readLength8() {
    this.logGroup('readLength8');
    let len = this.readU8();
    if (len & 0x80) {
      len = (len & 0x7f) << 8;
      len += this.readU8();
    }

    this.logDebug(`length: ${len}`);
    this.logGroupEnd();
    return len;
  }

  readLength16() {
    this.logGroup('readLength16');
    let len = this.readU16();
    if (len & 0x8000) {
      len = (len & 0x7fff) << 16;
      len += this.readU16();
    }

    this.logDebug(`length: ${len}`);
    this.logGroupEnd();
    return len;
  }

  // Unsigned Little Endian Base 128
  // Each byte has 7 bits allocated for data and 1 bit (the most significant bit, MSB) used as a continuation flag
  // The data bits (7 bits per byte) are concatenated to reconstruct the integer, with the least significant bits coming first (little-endian order)
  readUleb128(): number {
    let result = 0;
    let shift = 0;
    let byte;
    do {
      byte = this.readU8();
      result |= (byte & 0x7f) << shift;
      shift += 7;
    } while (byte >= 0x80);
    return result;
  }

  readUleb128Bigint(): bigint {
    let result = BigInt(0);
    let shift = 0;
    let byte;
    do {
      byte = this.readU8();
      result |= BigInt(byte & 0x7f) << BigInt(shift);
      shift += 7;
    } while (byte >= 0x80);
    return result;
  }

  // TODO(telkins)
  readLeb128(): number {
    let result = 0;
    let shift = 0;
    while (true) {
      const byte = this.buffer[this.cursor++];
      result |= (byte & 0x7f) << shift;
      if ((byte & 0x80) === 0) break;
      shift += 7;
    }
    return result;
  }

  readSizedInt(size: number): number {
    this.logGroup(`readSizedInt (${size} bytes)`);
    const cursor = this.cursor;
    let result: number;

    switch (size) {
      case 4:
        result =
          (this.buffer[cursor] & 0xff) |
          ((this.buffer[cursor + 1] & 0xff) << 8) |
          ((this.buffer[cursor + 2] & 0xff) << 16) |
          (this.buffer[cursor + 3] << 24);
        break;
      case 3:
        result =
          (this.buffer[cursor] & 0xff) | ((this.buffer[cursor + 1] & 0xff) << 8) | (this.buffer[cursor + 2] << 16);
        if (result & 0x800000) {
          result |= 0xff000000;
        }
        break;
      case 2:
        result = (this.buffer[cursor] & 0xff) | (this.buffer[cursor + 1] << 8);
        if (result & 0x8000) {
          result |= 0xffff0000;
        }
        break;
      case 1:
        result = this.buffer[cursor];
        if (result & 0x80) {
          result |= 0xffffff00;
        }
        break;
      default:
        throw new Error(`Invalid size ${size} for sized int at offset 0x${cursor.toString(16)}`);
    }

    this.cursor = cursor + size;
    this.logGroupEnd();
    return result;
  }

  readSizedUInt(size: number): number {
    this.logGroup(`readSizedUInt (${size} bytes)`);
    const cursor = this.cursor;
    let result: number;

    switch (size) {
      case 4:
        result =
          (this.buffer[cursor] & 0xff) |
          ((this.buffer[cursor + 1] & 0xff) << 8) |
          ((this.buffer[cursor + 2] & 0xff) << 16) |
          ((this.buffer[cursor + 3] & 0xff) << 24);
        break;
      case 3:
        result =
          (this.buffer[cursor] & 0xff) |
          ((this.buffer[cursor + 1] & 0xff) << 8) |
          ((this.buffer[cursor + 2] & 0xff) << 16);
        break;
      case 2:
        result = (this.buffer[cursor] & 0xff) | ((this.buffer[cursor + 1] & 0xff) << 8);
        break;
      case 1:
        result = this.buffer[cursor] & 0xff;
        break;
      default:
        throw new Error(`Invalid size ${size} for sized int at offset 0x${cursor.toString(16)}`);
    }

    this.cursor = cursor + size;
    this.logGroupEnd();
    return result >>> 0;
  }

  readSizedLong(size: number): bigint {
    this.logGroup(`readSizedLong (${size} bytes)`);
    let result = BigInt(0);
    if (size < 1 || size > 8) {
      throw new Error(`Invalid size ${size} for sized long at offset 0x${this.cursor.toString(16)}`);
    }

    // Reading bytes from the buffer and constructing the long (bigint)
    for (let i = 0; i < size; i++) {
      result |= BigInt(this.buffer[this.cursor + i]) << (BigInt(i) * BigInt(8));
    }

    this.cursor += size;
    this.logDebug(`Read Sized Long: ${result}`);
    this.logGroupEnd();
    return result;
  }

  readSizedFloat(size: number): number {
    this.logGroup(`readSizedFloat (${size} bytes)`);
    const bytes = Buffer.alloc(4);

    // zero extend tot the right
    this.buffer.copy(bytes, 4 - size, this.cursor, (this.cursor += size));

    this.logGroupEnd();
    return bytes.readFloatLE(0);
  }

  readSizedDouble(size: number): number {
    this.logGroup(`readSizedDouble (${size} bytes)`);
    const bytes = Buffer.alloc(8);

    // zero extend tot the right
    this.buffer.copy(bytes, 8 - size, this.cursor, (this.cursor += size));

    this.logGroupEnd();
    return bytes.readDoubleLE(0);
  }

  readStringWithLength(length: number): string {
    this.logGroup(`readString (${length} bytes)`);
    const str = this.buffer.toString('utf8', this.cursor, this.cursor + length).replace(/\0/g, '');
    this.cursor += length;
    this.logGroupEnd();
    return str;
  }

  readStringNullTerminated(): string {
    let end = this.cursor;
    while (this.buffer[end] !== 0) end++;
    const str = this.buffer.toString('utf8', this.cursor, end);
    this.cursor = end + 1;
    return str;
  }

  maybeReadStringNullTerminated(): string | undefined {
    const str = this.readStringNullTerminated();
    return str.length === 0 || str === '\u0001' ? undefined : str;
  }

  slice(length: number): Buffer {
    const result = this.buffer.slice(this.cursor, this.cursor + length);
    this.cursor += length;
    return result;
  }

  alignBuffer(alignment: number = 4): void {
    const currentPos = this.cursor;
    if (currentPos % alignment !== 0) {
      this.skip(alignment - (currentPos % alignment));
    }
  }

  logGroup(name: string) {
    if (this.debug) console.group(name);
  }

  logGroupEnd() {
    if (this.debug) console.groupEnd();
  }

  logDebug(message: string) {
    if (this.debug) console.debug(message);
  }
}

File: src/launchpad/parsers/buffer_wrapper.py
"""Wrapper for binary buffer parsing with cursor tracking and debugging."""

from __future__ import annotations

import struct
import types

from dataclasses import dataclass

from launchpad.utils.logging import get_logger

logger = get_logger(__name__)


@dataclass
class DebugLogContext:
    """Context manager for debug logging groups."""

    name: str

    def __enter__(self) -> None:
        """Enter debug group."""
        # logger.debug("=== %s ===", self.name)

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: types.TracebackType | None,
    ) -> None:
        """Exit debug group."""
        # logger.debug("=== End %s ===", self.name)


class BufferWrapper:
    """Wrapper for binary buffer parsing with cursor tracking and debugging."""

    def __init__(self, buffer: bytes) -> None:
        """Initialize buffer wrapper.

        Args:
            buffer: Raw bytes to parse
            debug: Whether to enable debug logging
        """
        self.buffer = buffer
        self.cursor = 0

    def seek(self, offset: int) -> None:
        """Set cursor position.

        Args:
            offset: New cursor position
        """
        self.cursor = offset

    def skip(self, length: int) -> None:
        """Skip bytes.

        Args:
            length: Number of bytes to skip
        """
        self.cursor += length

    def _debug_group(self, name: str) -> DebugLogContext:
        """Create debug logging context.

        Args:
            name: Name of the debug group

        Returns:
            Debug context manager
        """
        return DebugLogContext(name)

    def read_u8(self) -> int:
        """Read unsigned 8-bit integer."""
        with self._debug_group("read_u8"):
            # logger.debug(f"cursor: {self.cursor}")
            val = self.buffer[self.cursor]
            # logger.debug(f"value: {val}")
            self.cursor += 1
            return val

    def read_s8(self) -> int:
        """Read signed 8-bit integer."""
        with self._debug_group("read_s8"):
            # logger.debug(f"cursor: {self.cursor}")
            val = struct.unpack("<b", self.buffer[self.cursor : self.cursor + 1])[0]
            # logger.debug(f"value: {val}")
            self.cursor += 1
            return val  # type: ignore[no-any-return]

    def read_u16(self) -> int:
        """Read unsigned 16-bit integer (little-endian)."""
        with self._debug_group("read_u16"):
            # logger.debug(f"cursor: {self.cursor}")
            val = struct.unpack("<H", self.buffer[self.cursor : self.cursor + 2])[0]
            # logger.debug(f"value: {val}")
            self.cursor += 2
            return val  # type: ignore[no-any-return]

    def read_s32(self) -> int:
        """Read signed 32-bit integer (little-endian)."""
        with self._debug_group("read_s32"):
            # logger.debug(f"cursor: {self.cursor}")
            val = struct.unpack("<i", self.buffer[self.cursor : self.cursor + 4])[0]
            # logger.debug(f"value: {val}")
            self.cursor += 4
            return val  # type: ignore[no-any-return]

    def read_u32(self) -> int:
        """Read unsigned 32-bit integer (little-endian)."""
        with self._debug_group("read_u32"):
            # logger.debug(f"cursor: {self.cursor}")
            val = struct.unpack("<I", self.buffer[self.cursor : self.cursor + 4])[0]
            # logger.debug(f"value: {val} 0x{val:08x}")
            self.cursor += 4
            return val  # type: ignore[no-any-return]

    def read_u32be(self) -> int:
        """Read unsigned 32-bit integer (big-endian)."""
        with self._debug_group("read_u32be"):
            # logger.debug(f"cursor: {self.cursor}")
            val = struct.unpack(">I", self.buffer[self.cursor : self.cursor + 4])[0]
            # logger.debug(f"value: {val} 0x{val:08x}")
            self.cursor += 4
            return val  # type: ignore[no-any-return]

    def read_u64(self) -> int:
        """Read unsigned 64-bit integer (little-endian)."""
        with self._debug_group("read_u64"):
            # logger.debug(f"cursor: {self.cursor}")
            val = struct.unpack("<Q", self.buffer[self.cursor : self.cursor + 8])[0]
            # logger.debug(f"value: {val} 0x{val:016x}")
            self.cursor += 8
            return val  # type: ignore[no-any-return]

    def read_length8(self) -> int:
        """Read length-prefixed 8-bit integer."""
        with self._debug_group("read_length8"):
            length = self.read_u8()
            if length & 0x80:
                length = ((length & 0x7F) << 8) | self.read_u8()
            # logger.debug(f"length: {length}")
            return length

    def read_length16(self) -> int:
        """Read length-prefixed 16-bit integer."""
        with self._debug_group("read_length16"):
            length = self.read_u16()
            if length & 0x8000:
                length = ((length & 0x7FFF) << 16) | self.read_u16()
            # logger.debug(f"length: {length}")
            return length

    def read_uleb128(self) -> int:
        """Read unsigned LEB128 integer.

        Each byte has 7 bits allocated for data and 1 bit (MSB) used as a continuation flag.
        The data bits are concatenated in little-endian order.
        """
        result = 0
        shift = 0
        while True:
            byte = self.read_u8()
            result |= (byte & 0x7F) << shift
            if not (byte & 0x80):
                break
            shift += 7
        return result

    def read_leb128(self) -> int:
        """Read signed LEB128 integer."""
        result = 0
        shift = 0
        while True:
            byte = self.read_u8()
            result |= (byte & 0x7F) << shift
            if not (byte & 0x80):
                # Sign extend if necessary
                if shift < 32 and (byte & 0x40):
                    result |= ~((1 << shift) - 1)
                break
            shift += 7
        return result

    def read_sized_int(self, size: int) -> int:
        """Read signed integer of specified size.

        Args:
            size: Size in bytes (1-4)

        Returns:
            Signed integer value

        Raises:
            ValueError: If size is invalid
        """
        with self._debug_group(f"read_sized_int ({size} bytes)"):
            if not 1 <= size <= 4:
                raise ValueError(
                    f"Invalid size {size} for sized int at offset 0x{self.cursor:08x}"
                )

            # Read bytes and sign extend
            if size == 4:
                val = struct.unpack("<i", self.buffer[self.cursor : self.cursor + 4])[0]
            else:
                # Read as unsigned and sign extend
                val = 0
                for i in range(size):
                    val |= self.buffer[self.cursor + i] << (i * 8)
                # Sign extend
                if val & (1 << ((size * 8) - 1)):
                    val |= ~((1 << (size * 8)) - 1)

            self.cursor += size
            return val  # type: ignore[no-any-return]

    def read_sized_uint(self, size: int) -> int:
        """Read unsigned integer of specified size.

        Args:
            size: Size in bytes (1-4)

        Returns:
            Unsigned integer value

        Raises:
            ValueError: If size is invalid
        """
        with self._debug_group(f"read_sized_uint ({size} bytes)"):
            if not 1 <= size <= 4:
                raise ValueError(
                    f"Invalid size {size} for sized uint at offset 0x{self.cursor:08x}"
                )

            # Read bytes
            val = 0
            for i in range(size):
                val |= self.buffer[self.cursor + i] << (i * 8)

            self.cursor += size
            return val

    def read_sized_float(self, size: int) -> float:
        """Read float of specified size.

        Args:
            size: Size in bytes (1-4)

        Returns:
            Float value

        Raises:
            ValueError: If size is invalid
        """
        with self._debug_group(f"read_sized_float ({size} bytes)"):
            if not 1 <= size <= 4:
                raise ValueError(
                    f"Invalid size {size} for sized float at offset 0x{self.cursor:08x}"
                )

            # Zero extend to 4 bytes
            bytes_val = bytearray(4)
            bytes_val[4 - size :] = self.buffer[self.cursor : self.cursor + size]
            val = struct.unpack("<f", bytes_val)[0]

            self.cursor += size
            return val  # type: ignore[no-any-return]

    def read_sized_double(self, size: int) -> float:
        """Read double of specified size.

        Args:
            size: Size in bytes (1-8)

        Returns:
            Double value

        Raises:
            ValueError: If size is invalid
        """
        with self._debug_group(f"read_sized_double ({size} bytes)"):
            if not 1 <= size <= 8:
                raise ValueError(
                    f"Invalid size {size} for sized double at offset 0x{self.cursor:08x}"
                )

            # Zero extend to 8 bytes
            bytes_val = bytearray(8)
            bytes_val[8 - size :] = self.buffer[self.cursor : self.cursor + size]
            val = struct.unpack("<d", bytes_val)[0]

            self.cursor += size
            return val  # type: ignore[no-any-return]

    def read_string_with_length(self, length: int) -> str:
        """Read string of specified length.

        Args:
            length: String length in bytes

        Returns:
            Decoded string with null bytes removed
        """
        with self._debug_group(f"read_string ({length} bytes)"):
            val = (
                self.buffer[self.cursor : self.cursor + length]
                .decode("utf-8", errors="replace")
                .replace("\0", "")
            )
            self.cursor += length
            return val

    def read_string_null_terminated(self) -> str:
        """Read null-terminated string.

        Returns:
            Decoded string
        """
        end = self.cursor
        while end < len(self.buffer) and self.buffer[end] != 0:
            end += 1
        val = self.buffer[self.cursor : end].decode("utf-8", errors="replace")
        self.cursor = end + 1
        return val

    def maybe_read_string_null_terminated(self) -> str | None:
        """Read null-terminated string if it exists.

        Returns:
            Decoded string or None if empty or special marker
        """
        val = self.read_string_null_terminated()
        return None if not val or val == "\x01" else val

    def slice(self, length: int) -> bytes:
        """Get slice of buffer.

        Args:
            length: Length of slice

        Returns:
            Slice of buffer
        """
        val = self.buffer[self.cursor : self.cursor + length]
        self.cursor += length
        return val

    def align_buffer(self, alignment: int = 4) -> None:
        """Align cursor to specified boundary.

        Args:
            alignment: Alignment boundary (default: 4)
        """
        if self.cursor % alignment != 0:
            self.skip(alignment - (self.cursor % alignment))
